### PR TITLE
fix(bindings): More authentication service server name fixes.

### PR DESCRIPTION
**Pull request by pixlwave**
Entering `https://matrix-client.matrix.org/` failed with an error about being an invalid domain or IP address. This PR fixes that by
- Trimming any trailing slashes (and whitespace whilst we're here) when sanitising.
- Trying to build the client with a URL instead of throwing when server name fails (e.g. `https://company.com/matrix`).

Additionally it adds tests to both the sanitising (Rust) and the service (Swift).

### CODE REVIEW

File: bindings/matrix-sdk-ffi/src/authentication_service.rs [hunk 1]
```diff
@@ -126,26 +127,34 @@ impl AuthenticationService {
...
-        let server_name = matrix_sdk::sanitize_server_name(&server_name_or_homeserver_url)
-            .map_err(AuthenticationError::from)?;
-
-        builder = builder.server_name(server_name.to_string());
-
-        let client = builder
-            .build()
-            .or_else(|e| {
-                if !server_name_or_homeserver_url.starts_with("http://")
-                    && !server_name_or_homeserver_url.starts_with("http://")
```
Comment (pixlwave) [inline bindings/matrix-sdk-ffi/src/authentication_service.rs:139]: Whoops, these were both the same 🤦‍♂️ Fixed!

File: crates/matrix-sdk/tests/integration/client.rs [hunk 2]
```diff
@@ -630,3 +631,29 @@ fn serialize_session() {
         })
     );
 }
+
+#[test]
+fn sanitize_server_name() {
```
Comment (pixlwave) [inline crates/matrix-sdk/tests/integration/client.rs:636]: Should I create a new file for this? I tried adding it to `main.rs` but it didn't run.

**Comment by pixlwave** [comment]
Wow, impressed that it found a typo in my Swift 😆

File: bindings/matrix-sdk-ffi/src/authentication_service.rs [hunk 3]
```diff
@@ -126,26 +127,34 @@ impl AuthenticationService {
...
+        let result = matrix_sdk::sanitize_server_name(&server_name_or_homeserver_url);
+        match result {
+            Ok(server_name) => {
+                builder = builder.server_name(server_name.to_string());
+            }
+            Err(e) => {
+                // When the input isn't a valid server name check it is a URL.
+                // If this is the case, build the client with a homeserver URL.
+                if let Ok(_url) = Url::parse(&server_name_or_homeserver_url) {
+                    builder = builder.homeserver_url(server_name_or_homeserver_url.clone());
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/authentication_service.rs:140]: I think this shouldn't just be a regular fallback here, but rather integrated with the UI in some way (I think users should get some sort of warning that tells them about broken discovery so they can let their HS admin know).

File: crates/matrix-sdk/tests/integration/client.rs (hunk 2 above)
Comment (jplatte) [reply crates/matrix-sdk/tests/integration/client.rs:636]: This should be in `main.rs`, are you sure it didn't run? It's in a different piece of the output then (unit tests, not integration tests).

Comment (pixlwave) [reply crates/matrix-sdk/tests/integration/client.rs:636]: Hmmm ok, I'll give it another try 👀

Comment (pixlwave) [reply crates/matrix-sdk/tests/integration/client.rs]: Ok yes, worked fine this time. Not sure what my issue was before :)

Comment (pixlwave) [reply crates/matrix-sdk/tests/integration/client.rs]: Like so?

**Review by jplatte — APPROVED** [review]

File: crates/matrix-sdk/tests/integration/client.rs (hunk 2 above)
Comment (jplatte) [reply crates/matrix-sdk/tests/integration/client.rs]: Yes!

**Pull request merged**